008-dplyr-retain-and-exclude.Rmd

```diff
@@ -0,0 +1,583 @@
+```
+
+#### With `anti_join()`
+
+Savvy dplyr users might recognize that the above `filter()` could also be expressed as an anti join to avoid dealing with missing values:
+
+```{r}
+anti_join(
+  cars,
+  cars |> filter(class == "suv" | mpg < 15),
+  by = join_by(class, mpg)
+)
+```
+
+I'd argue that this is much "too clever" for the problem at hand.
+It also gets prohibitively slow as the size of the data increases because you are anti-joining across all columns at once.
+
+#### `tidyr::drop_na()`
+
+`tidyr::drop_na()` is a special case of `exclude(this | that)`.
+Because dplyr didn't have an "exclude rows" solution, years ago we added `tidyr::drop_na()` as a way to drop rows where *any* specified column is `NA`.
+The goal here is:
+
+> *Exclude* rows where `deceased` *or* `date` are missing.
+
+```{r}
+patients <- tibble(
+  deceased = c(0, NA, 0, 1, NA, 0),
+  date = c(2005, 2010, NA, 2020, 2010, 2000)
+)
+
+patients |> tidyr::drop_na(deceased, date)
+
+# Equivalent `filter()`s
+patients |> filter(!(is.na(deceased) | is.na(date)))
+
+patients |> filter(!if_any(c(deceased, date), is.na))
+```
+
+With `exclude()`, you can express this as sequential `exclude()` calls if you just have 2-3 columns to work with:
+
+```{r}
+patients |>
+  exclude(is.na(deceased)) |>
+  exclude(is.na(date))
+```
+
+Or, if you have many columns, you can use `if_any()` like with the `filter()` example above, but in a more readable form:
+
+```{r}
+patients |> exclude(if_any(c(deceased, date), is.na))
+```
+
+### Retaining rows using `this | that`
+
+We've talked a lot about *excluding* rows, but this tidyup also proposes a feature that helps with *retaining* rows when using conditions combined with `|` - `when_any()`.
+
+Our goal here is:
+
+> *Retain* rows for "US" and "CA" when the score is between 200-300, *or* rows for "PR" and "RU" when the score is between 100-200.
+
+```{r}
+countries <- tibble(
+  name = c("US", "CA", "PR", "RU", "US", NA, "CA", "PR"),
+  score = c(200, 100, 150, NA, 50, 100, 300, 250)
+)
+
+countries
+```
+
+Here's a `filter()` solution, note how we've introduced 3 boolean operators, `&`, `|`, and `()`, decreasing readability and increasing the mental gymnastics required to understand it:
+
+```{r}
+countries |>
+  filter(
+    (name %in% c("US", "CA") & between(score, 200, 300)) |
+      (name %in% c("PR", "RU") & between(score, 100, 200))
+  )
+```
+
+With `when_any()`, you specify `,` separated conditions like you're used to, but they get combined with `|` rather than `&`.
+This allows us to reduce the amount of boolean operators introduced down to just `&`, and it remains pretty readable:
+
+```{r}
+countries |>
+  retain(when_any(
+    name %in% c("US", "CA") & between(score, 200, 300),
+    name %in% c("PR", "RU") & between(score, 100, 200)
+  ))
+```
+
+We think this is a better solution than adding a `.combine = c("&", "|")` style argument to `retain()` itself, which feels less readable overall:
```

Comment:
Based on the function names, I'd call this `.when = c("all", "any")`. 

It feels a bit weird to me to introduce a `when_all()` that you shouldn't generally need to use?